File: solution/0025.Reverse Nodes in k-Group/Solution.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode reverseKGroup(ListNode head, int k) {
        if(head == null || k < 2) {
            return head;
        }
        int num = 0;
        ListNode pNode = head;
        ListNode lastNode = new ListNode(0);
        ListNode reNode = lastNode;
        lastNode.next = head;
        while (pNode != null) {
            num++;
            if(num >= k) {
                num = 0;
                ListNode tempNode = pNode.next;
                reverse(lastNode.next, k);
				// k 个节点的尾节点指向下一组的头节点
                lastNode.next.next = tempNode;	
				// 上一组的尾节点指向当前 k 个节点的头节点				
                tempNode = lastNode.next;				
                lastNode.next = pNode;
				
                lastNode = tempNode;
                pNode = lastNode.next;
            }
            else {
                pNode = pNode.next;
            }
        }
        return reNode.next;
    }

    private ListNode reverse(ListNode node, int i) {
        if(i <= 1 || node.next == null) {
            return node;
        }
        ListNode lastNode = reverse(node.next, i - 1);
        lastNode.next = node; 
        return node;
    }
}

File: solution/0025.Reverse Nodes in k-Group/Solution.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
 func reverseKGroup(head *ListNode, k int) *ListNode {
	start := head
	var h *ListNode = nil
	var t *ListNode = nil
	ok := false
	if h, t, ok = revert(start, k); !ok {
		return head
	}
	r := h
	p := t
	print(2, r)
	for ok {
		start = t.Next
		h, t, ok = revert(start, k)
		p.Next = h
		p = t
	}
	return r
}

func revert(head *ListNode, k int) (h, t *ListNode, ok bool) {
	c := head
	if !check(c, k) {
		return head, nil, false
	}

	t = head
	var p *ListNode = nil
	for k > 0 {
		temp := head
		head = head.Next
		temp.Next = p
		p = temp
		k--
	}
	h = p
	t.Next = head
	ok = true
	print(11, h)
	return
}

func check(head *ListNode, k int) bool {
	for k > 0 {
		if head == nil {
			return false
		}
		k--
		head = head.Next
	}
	return true
}